controllers/cloudflare_api.go

```diff
@@ -546,3 +547,267 @@ func (c *CloudflareAPI) InsertOrUpdateTXT(fqdn, txtId, dnsId string) error {
+	}
+
+	if exists {
+		newApp.ID = id
+		c.Log.Info("updating access application", "name", name, "id", id)
+		_, err := c.CloudflareClient.UpdateAccessApplication(ctx, accountId, newApp)
+		if err != nil {
+			c.Log.Error(err, "error updating access application", "name", name)
+			return err
+		}
+	} else {
+		c.Log.Info("creating access application", "name", name)
+		newApp.ID = ""
+		createdApp, err := c.CloudflareClient.CreateAccessApplication(ctx, accountId, newApp)
+		if err != nil {
+			c.Log.Error(err, "error creating access application", "name", name)
+			return err
+		}
+
+		// Set id, so that it can be used in policy creation
+		id = createdApp.ID
+	}
+
+	// Handle application policies
+	err = c.createAccessApplicationPolicies(accountId, id, config)
+	if err != nil {
+		c.Log.Error(err, "error creating access application policies", "name", name)
+		return err
+	}
+
+	c.Log.Info("access application reconciled successfully", "name", name, "existing", exists)
+	return nil
+}
+
+func (c *CloudflareAPI) DeleteAccessConfig(name string, config networkingv1alpha1.AccessConfig) error {
+	ctx := context.Background()
+	accountId := c.ValidAccountId
+
+	exists, id, err := c.getAccessApplicationIdByName(name)
+	if err != nil {
+		c.Log.Error(err, "failed retrieving application for account", "accountId", accountId, "app", name)
+	}
+
+	if exists {
+		c.Log.Info("deleting access application", "name", name, "id", id)
+		err := c.CloudflareClient.DeleteAccessApplication(ctx, accountId, id)
+		if err != nil {
+			c.Log.Error(err, "error deleting access application", "name", name)
+			return err
+		}
+	} else {
+		err := fmt.Errorf("application does not exist", "name", name, "id", id)
+		return err
+	}
+
+	c.Log.Info("access application deleted successfully", "name", name, "existing", exists)
+	return nil
+}
+
+func (c *CloudflareAPI) createAccessApplicationPolicies(accountId string, applicationId string, config networkingv1alpha1.AccessConfig) error {
+	ctx := context.Background()
+
+	// Check if there are any policies we need to delete
+	existingPolicies, _, err := c.CloudflareClient.AccessPolicies(ctx, accountId, applicationId, cloudflare.PaginationOptions{})
+	if err != nil {
+		c.Log.Error(err, "failed retrieving existing access policies for application", "applicationId", applicationId)
+	}
+
+	if len(existingPolicies) > 0 {
+		for _, policy := range existingPolicies {
+
+			// We need to delete all existing policies
+			if len(config.AccessPolicies) == 0 {
+				c.Log.Info("deleting access policy for application", "applicationId", applicationId, "policyId", policy.ID)
+
+				err := c.CloudflareClient.DeleteAccessPolicy(ctx, accountId, applicationId, policy.ID)
+				if err != nil {
+					c.Log.Error(err, "error deleting access policy for application", "applicationId", applicationId, "policyId", policy.ID)
+					return err
+				}
+			}
+
+			// Check if the policy is still required
+			required := false
```

Comment:
This function seems convoluted. Let me phrase what I understand this function should be doing and you can correct me.

Given the set of existing access policies on the application, and the set of access policies defined in the CRD, you are trying to reconcile them. But there are a couple of problems. Does having the same name of the policy guarantee that the policy wasn't changed? Do we also create access groups here or just refer to them by name?

If possible, I'd like to see this split up into smaller functions and sync all the rules regardless of what was present, unless there was a way you could verify what is set online matches the local config.

Reply:
This one was a bit tricky.
What I am doing here is checking by name if there are any policies in the application, which are not defined in crd.
If yes, we delete them.

Later on we loop trough policies defined in crd and update them.

Cases I tested:
- Policy by same name is changed manually, it should be updated to be matching crd configuration - works
- Policy manually added to the ui, should be deleted - works
- Policy should be deleted once gone from crd - works

Reply:
does this makes sense?